src/ompl/base/samplers/DeterministicStateSampler.h

```diff
@@ -0,0 +1,172 @@
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Leonard Bruns */
+
+#ifndef OMPL_BASE_DETERMINISTIC_STATE_SAMPLER_
+#define OMPL_BASE_DETERMINISTIC_STATE_SAMPLER_
+
+#include "ompl/base/StateSpace.h"
+#include "ompl/base/StateSampler.h"
+#include "ompl/base/samplers/deterministic/DeterministicSequence.h"
+
+#include <memory>
+
+namespace ompl
+{
+    namespace base
+    {
+        /**
+       @anchor DeterministicStateSampler
+       @par Short description
+       \ref DeterministicStateSampler Implementation of a deterministic state sampler. The
+       implementation allows to load and draw samples from a precomputed sequence or from the Halton Sequence,
+       @par External documentation
+       Dispertio: Optimal Sampling For Safe Deterministic Motion Planning
+       L Palmieri, L Bruns, M Meurer, KO Arras - IEEE Robotics and Automation Letters, 2019
+       DOI: [10.1109/LRA.2019.2958525](https://ieeexplore.ieee.org/abstract/document/8928532)<br>
+       [[PDF]](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=8928532)
+    */
+
+        /** \brief An abstract class for the concept of using deterministic sampling sequences
+        to decrease the dispersion of the samples. */
+        class DeterministicStateSampler : public StateSampler
+        {
+        public:
+            enum DeterministicSamplerType
+            {
+                HALTON
+            };
+
+            /** \brief Constructor, which creates the sequence internally based on the specified sequence type.
+            Uses the default constructor for the sequence.*/
+            DeterministicStateSampler(const StateSpace *space,
+                                      DeterministicSamplerType type = DeterministicSamplerType::HALTON);
+            /** \brief Constructor that takes a pointer to a DeterministicSequence and uses that object instead
+            of its own. This can be used to apply non default options to the deterministic sequence.*/
+            DeterministicStateSampler(const StateSpace *space, std::shared_ptr<DeterministicSequence> sequence_ptr);
+
+            void sampleUniform(State *state) override = 0;
```

Comment:
You should provide a default implementation that uses StateSpace::copyFromReals to copy the doubles from an std::vector into the state. You only need to override this behavior if you need to scale the coordinates.

Reply:
Hi Mark, can you comment a bit more on this? Or maybe provide an example? Not clear what you meant.

Reply:
This method need not be pure virtual. It can provide a default implementation to copy the next deterministic sample, stored in a std::vector<double>, to an abstract State*. The function StateSpace::copyFromReals is designed to do exactly this. It knows how to map the different vector values to memory locations inside the State*. This way you don't have to implement this method for every possible state space.

Reply:
Implemented